### PR TITLE
examples : add tokenization tests and refactor codes

### LINKED ISSUE #170: gpt_tokenize : bug in tokenization and incapable of double-byte languages
There is a bug in gpt_tokenize function in examples/common.cpp 

Followings are comparison with a huggingface tokenizer, 
```
# huggingface
>>> from transformers import AutoTokenizer, AutoModelForCausalLM
>>> tokenizer = AutoTokenizer.from_pretrained("EleutherAI/gpt-neox-20b")
>>> tokenizer.tokenize("ableable")
['able', 'able']
```

```
# ggml
main: number of tokens in prompt = 5
main: token[0] =  21762, able
main: token[1] =     68, a
main: token[2] =     69, b
main: token[3] =     79, l
main: token[4] =     72, e
```

Example above is due to this line https://github.com/ggerganov/ggml/blob/2a75bd43113702fe096c75408b300d106be90a58/examples/common.cpp#L269

Furthermore, current implementation does not consider double-byte characters such as Korean, Japanese, Chinese. I'm wondering whether it would be ideal to modify the gpt_tokenize function to be able to handle double-byte characters, or the repo aims on single character languages only as English and most roman languages are single-byte.

### PR DESCRIPTION
**Pull request by jaeminSon**
(An attempt to fix #170)

examples : test tokenization + support utf-8 and multi-byte encodings + refactor gpt_tokenize + refactor convert-h5-to-ggml.py

### CODE REVIEW

File: examples/common.cpp [hunk 1]
```diff
@@ -212,34 +222,22 @@ void gpt_vocab::add_special_token(const std::string & token) {
...
 std::vector<gpt_vocab::id> gpt_tokenize(const gpt_vocab & vocab, const std::string & text) {
     std::vector<std::string> words;
 
     // Convert input to utf-8
-    std::string utf8conv;
-    for (int w = 0; w < text.size(); w++) {
-        append_utf8( uint8_t(text[w]), utf8conv);
-    }
+    std::wstring text_multibytes = convert_to_wstring(text);
+    std::string utf8conv = convert_to_utf8(text_multibytes);
```
Comment (klosax) [inline examples/common.cpp:240]: I reverted this in pr #184 as it breaks other models, see comment in pr #179 . I think each model will have to handle the encoding of its tokens.

File: examples/common.cpp [hunk 2]
```diff
@@ -8,6 +8,16 @@
...
+#include <locale>
+#include <codecvt>
+#include <sstream>
+#include <filesystem>
+
+#ifdef __APPLE__
+namespace fs = std::__fs::filesystem;
+#else
+namespace fs = std::filesystem;
+#endif
```
Comment (ggerganov) [inline examples/common.cpp:20]: Do not use `#include <filesystem>`. Use `<fstream>`

File: examples/common.cpp [hunk 3]
```diff
@@ -271,43 +255,136 @@ std::vector<gpt_vocab::id> gpt_tokenize(const gpt_vocab & vocab, const std::stri
...
+
     return tokens;
 }
 
+bool are_strings_equal(const std::vector<std::string>& vec1, const std::vector<std::string>& vec2) {
+    if (vec1.size() != vec2.size()) return false;
+
+    for (size_t i = 0; i < vec1.size(); ++i) 
+        if (vec1[i] != vec2[i]) return false;
+
```
Comment (ggerganov) [inline examples/common.cpp:287]:
```suggestion
    for (size_t i = 0; i < vec1.size(); ++i) {
        if (vec1[i] != vec2[i]) return false;
    }

```

Fix this in other places too

**Review by ggerganov — CHANGES_REQUESTED** [review]

File: examples/prompts/tokenize_huggingface.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+                "openai/whisper-tiny"  # whisper (base, large, large-v2 share the same tokenizer)
+                ]
+
+repo2ggml = {"databricks/dolly-v2-3b": "dolly-v2",
+             "gpt2": "gpt-2",
+             "uer/gpt2-chinese-cluecorpussmall": "gpt-2-chinese",
+             "EleutherAI/gpt-j-6b": "gpt-j",
+             "EleutherAI/gpt-neox-20b": "gpt-neox",
+             "EleutherAI/polyglot-ko-1.3b": "polyglot-ko",
+             "rinna/japanese-gpt-neox-3.6b": "japanese-gpt-neox",
```
Comment (ggerganov) [inline examples/prompts/tokenize_huggingface.py:25]:
```suggestion
             "rinna/japanese-gpt-neox-3.6b": "gpt-neox-japanese",
```

For consistency in the file names, always put the language as suffix

File: examples/common.cpp [hunk 5]
```diff
@@ -271,43 +251,149 @@ std::vector<gpt_vocab::id> gpt_tokenize(const gpt_vocab & vocab, const std::stri
...
+                if (fname.find(t_name) != std::string::npos) {
+                    return dir_test + "/" + t_fname;
+                }
+            }
+        }
+        closedir(dir);
+    }
+    
+    return ""; // empty string if test file not found
+}
```
Comment (ggerganov) [inline examples/common.cpp:330]: I think this entire function is not needed.
Instead, let's explicitly provide the path to the test file in each example

**Review by ggerganov — APPROVED** [review]
Great job 👍 

After resolving the merge conflicts, we can merge it

**Pull request merged**